src/RageSoundMixBuffer.cpp

```diff
@@ -31,19 +31,36 @@ void RageSoundMixBuffer::SetWriteOffset( int iOffset )
 	m_iOffset = iOffset;
 }
 
-void RageSoundMixBuffer::Extend( unsigned iSamples )
+void RageSoundMixBuffer::Extend(unsigned iSamples)
 {
-	const unsigned realsize = iSamples+m_iOffset;
-	if( m_iBufSize < realsize )
+	const std::uint64_t m_nBufSize = static_cast<std::uint64_t>(iSamples) + m_iOffset;
+	if (m_iBufSize < m_nBufSize)
 	{
-		m_pMixbuf = (float *) realloc( m_pMixbuf, sizeof(float) * realsize );
-		m_iBufSize = realsize;
+		m_pMixbufBackup = static_cast<float*>(realloc(m_pMixbufBackup, sizeof(float) * m_iBufSize));
+		if (m_pMixbufBackup != nullptr && m_pMixbuf != nullptr)
+		{
+			memcpy(m_pMixbufBackup, m_pMixbuf, m_iBufSize * sizeof(float));
+		}
+
+		float *m_pMixbufNew = static_cast<float*>(realloc(m_pMixbuf, sizeof(float) * m_nBufSize));
+		if (m_pMixbufNew == nullptr)
```

Comment:
This indicates that `realloc()` has failed. The code below tries to recover, but the whole point of `Extend()` is to change the size of the buffer. Other code which calls `Extend()` will assume that the buffer size has changed after this call and might cause memory corruption. I think we should just crash in this case.

Reply:
Isn't it better to have a buffer that's slightly too small for a few seconds so the game keeps running, instead of crashing? Since the buffer is exhausted and refilled very quickly, this way gameplay isn't affected, at worse the sound glitches out for a moment, but the game can keep running smoothly after that.

Reply:
The problem is that if the buffer is too small and we try to write too much data, this will cause a memory corruption (and it might be worse than just audio glitching out). And if you look for example in `RageSoundMixBuffer::write()` function, it first calls `Extend()`, and then assumes that buffer was resized properly without checking the size. So to avoid the memory corruption, we'd have to check the actual size after the call to `Extend()` for all places where it's used. Then it's possible to recover gracefully (possibly with just an audio glitch).

But why would `realloc()` even fail? Wouldn't that indicate that we're quite low on memory? Seems like crash is imminent anyway.

Reply:
Again, it should never fail to reallocate the memory, but it can and does happen and things like OS and audio driver will affect this. For example Windows is already incredibly resilient (but not impervious to this). But on Ubuntu it's extremely easy to mess up the audio buffer by starting and stopping OBS while using PipeWire with ALSA to capture gameplay.

A few other things... m_iOffset should never be negative, but it _could_ be. It's not likely enough to use `std::abs(m_iOffset)` in the `m_nBufSize` calculation, but just in case it does happen, having the backup buffer is going to be good then.

The original code was storing the sum of two `unsigned`'s in an `unsigned` without checking that it doesn't overflow. If it ever did, the buffer could end up being extremely small.

I've tested this code quite heavily in the last day or so and I've found that it does hold already recover gracefully, with just an audio glitch, in its current state.